core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -377,9 +376,9 @@ class LogSegment(val log: FileRecords,
   def timeWaitedForRoll(now: Long, messageTimestamp: Long) : Long = {
     // Load the timestamp of the first message into memory
     if (rollingBasedTimestamp.isEmpty) {
-      val iter = log.shallowEntries.iterator()
+      val iter = log.batches.iterator()
       if (iter.hasNext)
-        rollingBasedTimestamp = Some(iter.next().record.timestamp)
+        rollingBasedTimestamp = Some(iter.next().maxTimestamp)
```

Comment:
The behavior is not actually different. Recall that this is "shallow" iteration in the old code. The timestamp we are accessing is actually the max timestamp when the record is viewed as a batch. In the uncompressed case, the "batch" has just a single record so the timestamp is the same as its max timestamp. In the compressed case, the timestamp is the max timestamp of all records in the batch. We preserve the old behavior and adopt the old compressed behavior for the new message format. I can update the comment to clarify this.